include/revng-c/RestructureCFG/CycleEquivalenceAnalysis.h

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+//
+// Copyright rev.ng Labs Srl. See LICENSE.md for details.
+//
+
+#include <map>
+
+#include "llvm/IR/CFG.h"
+#include "llvm/Pass.h"
+
+#include "revng/ADT/GenericGraph.h"
+
+#include "CycleEquivalenceClass.h"
+
+// `struct` used to represent the unique incremental `Index` that is
+// associated with each undirected edge in the `GenericGraph`
+struct EdgeIndex {
+private:
+  size_t Index = 0;
+
+public:
+  EdgeIndex(size_t Index) { this->Index = Index; }
+  size_t getValue() const { return Index; }
+  bool operator<(const EdgeIndex &Rhs) const { return Index < Rhs.getValue(); }
+  bool operator==(const EdgeIndex &Rhs) const {
+    return Index == Rhs.getValue();
+  }
+};
+
+template<class GraphT, class GT = llvm::GraphTraits<GraphT>>
+class CycleEquivalenceAnalysis {
+public:
+  using NodeT = typename GT::NodeRef;
+
+private:
+  struct BlockBidirectionalNode {
+    BlockBidirectionalNode(NodeT Block) : Block(Block) {}
+    NodeT Block;
+    NodeT getBlock() { return Block; }
+    llvm::StringRef getName() const { return Block->getName(); }
+  };
+
+public:
+  using EdgeLabel = EdgeIndex;
+  using BlockNode = BidirectionalNode<BlockBidirectionalNode, EdgeLabel>;
+  using BlockGraph = GenericGraph<BlockNode>;
+  using EdgeDescriptor = revng::detail::EdgeDescriptor<BlockNode *>;
+  using EdgesVectorType = llvm::SmallVector<EdgeDescriptor>;
+  using BracketDescriptor = std::pair<EdgeDescriptor, size_t>;
+
+private:
+  llvm::SmallVector<CycleEquivalenceClass<NodeT>, 1> CycleEquivalenceClasses;
+
+  std::map<BracketDescriptor, size_t> BracketToClass;
+  std::map<size_t, BracketDescriptor> ClassToBracketDescriptor;
+
+  std::map<BlockNode *, size_t> NodeDFSNumMap;
+  std::map<size_t, BlockNode *> DFSNumNodeMap;
```

Comment:
This can be a `std::vector<BlockNode *>` or a `SmallVector<BlockNode *>`.

Reply:
This field is now not a part of the `CycleEquivalenceAnalysis` class anymore, but is still needed here https://github.com/revng/revng-c/blob/323b04da3a0979143e3207968b19e7142893109e/lib/RestructureCFG/CycleEquivalenceAnalysis.cpp#L169 in the `insertCappingBackedges` method.

This is due to the fact that the logic which computes the target of the capping backedge (when we need to insert one), uses merely the DFS index of each node in the computations. Therefore, at the end of the _routine_, we need to go back from the index to the node itself. I now populate this map (which is local to the method) opportunistically each time I have a node available (and can therefore access its DFSNum). The alternative would be to keep a pair of DFSNum and the corresponding node during the algorithm, but that solution would be strictly more memory consuming than the current approach.

Reply:
If you keep it as a local variable like you do now it's fine